### DISCUSSION TITLE
How to avoid triggering compilation every time with `pmap(closure)` ? (vmap ok)

### DISCUSSION THREAD

**Discussion by marius311** [jax-ml/jax#11023] [Q&A] [answered]
I've got the following code pattern:

```python
class MyProblem():

    @partial(jax.jit, static_argnums=0)
    def f(self, x, y):
        print("compiling")
        return x + y

    def solve(self, xs, y):
        def calls_f(x):
            return self.f(x, y)
        return jax.pmap(calls_f)(xs)

prob = MyProblem()

prob.solve(jnp.array(range(4)), 0)
```

This seems to trigger compilation every time I run the last line. If I replace with `vmap` it only compiles once. A simplified version without the `calls_f` closure, I'm able to fix it with Equinox's `filter_pmap`, but that doesn't work here. 

I'm wondering how I can rewrite this to avoid the recompilation? Is `pmap` just fundamentally incompatible with closures? Thanks!

**Comment by YouJiacheng** [comment]
Hi. - In fact `pmap` will compile the function!
see the note [here](https://jax.readthedocs.io/en/latest/_autosummary/jax.pmap.html).
You can try this:
```python
class MyProblem():
    # jit here is unnecessary
    def f(self, x, y):
        print("compiling")
        return x + y

    def solve(self, xs, y):
        return jax.pmap(self.f, in_axes=(0, None))(xs, y)

prob = MyProblem()

prob.solve(jnp.array(range(4)), 0)
```

**Reply by marius311** [reply]
Thanks, yea saw that and figured that was the cause. But is there not any way to do it without rewriting the structure of my code? (which in my real case would hurt the logic / code-clarity by a lot more than here). I guess it goes back to my question about is `pmap` (or I guess really `jit`) just incompatible with closures?

**Reply by YouJiacheng** [reply]
I try to minimize the modification for complex code, see my new answer.

**Comment by YouJiacheng** [comment]
@marius311 I think `jit` is "incompatible" with closure(with new captured value every time).
But I believe there is a workaround:
```python
from jax.tree_util import Partial # note: not functools.partial
def call_fun(f, *args, **kwargs):
    return f(*args, **kwargs)
class MyProblem():

    @partial(jax.jit, static_argnums=0)
    def f(self, x, y):
        print("compiling")
        return x + y

    def calls_f_open(self, x, y):
        return self.f(x, y)

    def solve(self, xs, y):
        calls_f = Partial(self.calls_f_open, y=y)
        # now the "closure" is a Pytree, with self.calls_f_open as aux_data which has same hash for same self
        return jax.pmap(call_fun, in_axes=(None, 0))(calls_f, xs)

prob = MyProblem()

prob.solve(jnp.array(range(4)), 0)
```

**Reply by marius311** [reply]
Thanks, appreciate the effort looking into this. This does work, but ultimtately it still is just getting rid of the closure right? Which makes me explicilty specify that `calls_f ` (`calls_f_open` in this case) depends on `y`, rather than just letting it close over `y` which I suppose is the part I really want.

Anyway, this is making me realize an even more reduced example that doesnt involve `pmap` is just:

```python
@jax.jit
def f(x):
    print("compiling f")
    return x

def g(x):
    @jax.jit
    def calls_f(x):
        print("compiling calls_f")
        return f(x)
    return calls_f(x)

g(0) # compiles f every time
```

I would understand why `calls_f` is compiled every time, but I guess my questions boils down to why also `f`? 

To hazard a guess, I suppose that to compile, Jax always just traces through everything? It never uses cached JIT'ed functions which might otherwise be present from a previous compile? (but in theory it could?)

**Reply by sharadmv** [reply]
Yep, you're exactly right. When you call `g` you're defining a new `calls_f` every time. Nested `jit`-ted functions are just traced through and we don't hit any caches. If you hoist `calls_f` outside of `g`, then we won't be defining a new one every time and the compiled `calls_f` would be cached.

**Reply by marius311** [reply]
Got it, thanks. Although to be clear,

> When you call g you're defining a new calls_f every time

would explain why you see `compiling calls_f` every time, but might not explain why you see `compiling f` every time too, because in theory I guess that _could_ be cached (or maybe not due to some technical limitation, I'm definitely pushing the bounds of my Python/JAX knowledge here).

**Reply by YouJiacheng** [reply]
@marius311 JAX use hash of the function(closure) as identifier. And `g` or `solve` create a brand new closure with different hash value(and should be different for correctness if `y` change). Thus, every time it is a new function for JAX, and need to be compiled.

**Reply by marius311** [reply]
Right, but `f` is the same every time so the explanation for why _it_ gets recompiled every time too I think is just that Jax doesn't cache inner compiled functions.

**Reply by YouJiacheng** [reply]
@marius311 Oh - For `f`, JAX currently doesn't adopt sub-jaxpr or XLA sub-graph IIUC, i.e. always tracing inner function and inline it. Tracing can be avoided, but if JAX still inline the sub-jaxpr, compiling cost won't be improved significantly.

**Reply by YouJiacheng** [reply]
As @sharadmv said, "Nested `jit`-ted functions are just traced through", i.e. the inner `jit` is ignored, at least for now.

**Reply by YouJiacheng** [reply]
Actually there is an option `inline` for `jit`, but I think it is not available since JAX always inline(actually trace through) `jit`-ted sub functions.

**Reply by marius311** [reply]
I'd be curious if you could say more about "tracing can be avoided"? In my mental model, tracing is synonymous with "did the print statement get shown" and it seems like we're unable to remove the repeated print statement while also keeping the closure in my example?

**Reply by sharadmv** [reply]
Check out Matt's comments in this thread: https://github.com/google/jax/issues/7155

There's also exploratory work in avoiding the retracing (https://github.com/google/jax/pull/9181) but at the end of the day it's an optimization that won't drastically change the compiled program (it may inline some functions).

**Reply by YouJiacheng** [reply]
I mean if we modify JAX current implementation(not too much), we can use cached jaxpr of `f` and avoid tracing it again. But it is not compelling and could be misleading since compiling usually takes more time than tracing.

**Reply by marius311** [reply]
Got it, thanks both for all the replies, I think that answers everything for me.

**Comment by marius311** [comment, marked as answer]
To summarize the answer from the very helpful discussion:

> "Nested jit-ted functions are just traced through", i.e. the inner jit is ignored, at least for now.

And [this](https://github.com/google/jax/issues/7155#issuecomment-872319315) comment also explains it. Workarounds here include dont use a closure, or JIT the entire `solve` function if you can.
